source/slang/ir-sccp.cpp

```diff
@@ -0,0 +1,933 @@
+            // behavior, so we can freely pick a value for it on a whim.
+            None,
+
+            // The `Constant` flavor represents an instuction that we
+            // have only ever seen produce a single, fixed value. It's
+            // `value` field will hold that constant value.
+            Constant,
+
+            // The `Any` flavor represents an instruction that might produce
+            // different values at runtime, so we go ahead and approximate
+            // this as it potentially yielding any value whatsoever. A
+            // more precise analysis could use sets or intervals of values,
+            // but for SCCP anything that could take on more than 1 value
+            // at runtime is assumed to be able to take on *any* value.
+            Any,
+        };
+
+        // The flavor of this value (`None`, `Constant`, or `Any`)
+        Flavor  flavor;
+
+        // If this is a `Constant` lattice value, then this field
+        // points to the IR instruction that defines the actual constant value.
+        // For all other flavors it should be null.
+        IRInst* value = nullptr;
+
+        // For convenience, we define `static` factory functions to
+        // produce values of each of the flavors.
+
+        static LatticeVal getNone()
+        {
+            LatticeVal result;
+            result.flavor = Flavor::None;
+            return result;
+        }
+
+        static LatticeVal getAny()
+        {
+            LatticeVal result;
+            result.flavor = Flavor::Any;
+            return result;
+        }
+
+        static LatticeVal getConstant(IRInst* value)
+        {
+            LatticeVal result;
+            result.flavor = Flavor::Constant;
+            result.value = value;
+            return result;
+        }
+
+        // We also need to be able to test if two lattice
+        // values are equal, so that we can avoid updating
+        // downstream dependencies if our knowledge about
+        // an instruction hasn't actually changed.
+        //
+        bool operator==(LatticeVal const& that)
+        {
+            return this->flavor == that.flavor
+                && this->value == that.value;
+        }
+
+        bool operator!=(LatticeVal const& that)
+        {
+            return !( *this == that );
+        }
+    };
+
+    // If we imagine a variable (actually an SSA phi node...) that
+    // might be assigned lattice value A at one point in the code,
+    // and lattice value B at another point, we need a way to
+    // combine these to form our knowledge of the possible value(s)
+    // for the variable.
+    //
+    // In terms of computation on a lattice, we want the "meet"
+    // operation, which computes the lower bound on what we know.
+    // If we interpret our lattice values as sets, then we are
+    // trying to compute the union.
+    //
+    LatticeVal meet(LatticeVal const& left, LatticeVal const& right)
```

Comment:
If this is called a lot is seems ripe for table. There are so little combinations you can even store the table in an integer and extract with shift and mask. Maybe premature optimization, and the root of all evil too.